File: platforms/android/java/app/msime/client/dictionary/CustomKeyboardSkin.java
package app.msime.client;

import java.util.Base64;
import java.util.Arrays;
import java.util.Locale;
import org.json.JSONObject;

/** Bounded Android view of Apple's current custom touch-keyboard design. */
public final class CustomKeyboardSkin {
    private int background = 0xE8F0EB;
    private int keyBackground = 0xFFFFFF;
    private int keyForeground = 0x17251D;
    private int accent = 0x185C47;
    private int actionBackground = 0x185C47;
    private double cornerRadius = 8;
    private double borderWidth;
    private double shadow;
    private int pattern;
    private boolean monospaced;
    private String keyShape = "rounded";
    private String keyMaterial = "flat";
    private double keyOpacity = 1;
    private Integer gradientEnd;
    private boolean gradientHorizontal;
    private double patternOpacity = .15;
    private Integer customBorderColor;
    private byte[] photo;
    private double photoShade = .25;
    private double photoPosition = .5;

    private CustomKeyboardSkin() {}

    public static CustomKeyboardSkin defaults() { return new CustomKeyboardSkin(); }

    public static CustomKeyboardSkin from(JSONObject object) {
        CustomKeyboardSkin value = defaults();
        if (object == null) return value;
        value.background = color(object, "background", value.background);
        value.keyBackground = color(object, "keyBackground", value.keyBackground);
        value.keyForeground = color(object, "keyForeground", value.keyForeground);
        value.accent = color(object, "accent", value.accent);
        value.actionBackground = color(object, "actionBackground", value.actionBackground);
        value.cornerRadius = bounded(object.optDouble("cornerRadius", value.cornerRadius), 0, 20, 8);
        value.borderWidth = bounded(object.optDouble("borderWidth", 0), 0, 2, 0);
        value.shadow = bounded(object.optDouble("shadow", 0), 0, .4, 0);
        value.pattern = Math.max(0, Math.min(3, object.optInt("pattern", 0)));
        value.monospaced = object.optBoolean("monospaced", false);
        value.keyShape = oneOf(object.optString("keyShape", "rounded"),
            "rounded", "capsule", "ticket", "pebble");
        value.keyMaterial = oneOf(object.optString("keyMaterial", "flat"),
            "flat", "raised", "glass", "paper");
        value.keyOpacity = bounded(object.optDouble("keyOpacity", 1), .25, 1, 1);
        if (object.has("gradientEnd") && !object.isNull("gradientEnd"))
            value.gradientEnd = color(object, "gradientEnd", value.background);
        value.gradientHorizontal = object.optBoolean("gradientHorizontal", false);
        value.patternOpacity = bounded(object.optDouble("patternOpacity", .15), 0, .5, .15);
        if (object.has("customBorderColor") && !object.isNull("customBorderColor"))
            value.customBorderColor = color(object, "customBorderColor", value.accent);
        value.photoShade = bounded(object.optDouble("photoShade", .25), 0, .8, .25);
        value.photoPosition = bounded(object.optDouble("photoPosition", .5), 0, 1, .5);
        value.photo = photo(object.optString("photo", ""));
        return value;
    }

    static CustomKeyboardSkin fixture(int background, int keyBackground, int keyForeground,
            int accent, int actionBackground, double cornerRadius, double borderWidth,
            double shadow, int pattern, boolean monospaced, String keyShape,
            String keyMaterial, double keyOpacity, Integer gradientEnd,
            boolean gradientHorizontal, double patternOpacity, Integer customBorderColor,
            byte[] photo, double photoShade, double photoPosition) {
        CustomKeyboardSkin value = defaults();
        value.background = background & 0xFFFFFF;
        value.keyBackground = keyBackground & 0xFFFFFF;
        value.keyForeground = keyForeground & 0xFFFFFF;
        value.accent = accent & 0xFFFFFF;
        value.actionBackground = actionBackground & 0xFFFFFF;
        value.cornerRadius = bounded(cornerRadius, 0, 20, 8);
        value.borderWidth = bounded(borderWidth, 0, 2, 0);
        value.shadow = bounded(shadow, 0, .4, 0);
        value.pattern = Math.max(0, Math.min(3, pattern));
        value.monospaced = monospaced;
        value.keyShape = oneOf(keyShape, "rounded", "capsule", "ticket", "pebble");
        value.keyMaterial = oneOf(keyMaterial, "flat", "raised", "glass", "paper");
        value.keyOpacity = bounded(keyOpacity, .25, 1, 1);
        value.gradientEnd = gradientEnd == null ? null : gradientEnd & 0xFFFFFF;
        value.gradientHorizontal = gradientHorizontal;
        value.patternOpacity = bounded(patternOpacity, 0, .5, .15);
        value.customBorderColor = customBorderColor == null ? null : customBorderColor & 0xFFFFFF;
        value.photo = photo != null && photo.length <= 512_000 && supportedPhoto(photo)
            ? photo.clone() : null;
        value.photoShade = bounded(photoShade, 0, .8, .25);
        value.photoPosition = bounded(photoPosition, 0, 1, .5);
        return value;
    }

    private static int color(JSONObject object, String key, int fallback) {
        return (int) object.optLong(key, fallback) & 0xFFFFFF;
    }

    private static double bounded(double value, double minimum, double maximum, double fallback) {
        return Double.isFinite(value) ? Math.max(minimum, Math.min(maximum, value)) : fallback;
    }

    private static String oneOf(String value, String first, String second, String third, String fourth) {
        if (second.equals(value) || third.equals(value) || fourth.equals(value)) return value;
        return first;
    }

    private static byte[] photo(String encoded) {
        if (encoded.isEmpty() || encoded.length() > 682_668) return null;
        try {
            byte[] bytes = Base64.getDecoder().decode(encoded);
            return bytes.length <= 512_000 && supportedPhoto(bytes) ? bytes : null;
        } catch (IllegalArgumentException error) {
            return null;
        }
    }

    private static boolean supportedPhoto(byte[] bytes) {
        if (bytes.length >= 3 && (bytes[0] & 0xFF) == 0xFF
                && (bytes[1] & 0xFF) == 0xD8 && (bytes[2] & 0xFF) == 0xFF) return true;
        if (starts(bytes, new byte[] {(byte) 0x89, 'P', 'N', 'G', 13, 10, 26, 10})) return true;
        if (starts(bytes, "GIF87a".getBytes(java.nio.charset.StandardCharsets.US_ASCII))
                || starts(bytes, "GIF89a".getBytes(java.nio.charset.StandardCharsets.US_ASCII))) return true;
        return bytes.length >= 12 && starts(bytes, new byte[] {'R', 'I', 'F', 'F'})
            && bytes[8] == 'W' && bytes[9] == 'E' && bytes[10] == 'B' && bytes[11] == 'P';
    }

    private static boolean starts(byte[] bytes, byte[] prefix) {
        if (bytes.length < prefix.length) return false;
        for (int index = 0; index < prefix.length; index++)
            if (bytes[index] != prefix[index]) return false;
        return true;
    }

    private static String hex(int value) {
        return String.format(Locale.ROOT, "#%06X", value & 0xFFFFFF);
    }

    public String background() { return hex(background); }
    public String keyBackground() { return hex(keyBackground); }
    public String keyForeground() { return hex(keyForeground); }
    public String accent() { return hex(accent); }
    public String actionBackground() { return hex(actionBackground); }
    public String actionForeground() { return luminance(actionBackground) > .179 ? "#000000" : "#FFFFFF"; }
    public double cornerRadius() { return cornerRadius; }
    public double borderWidth() { return borderWidth; }
    public double shadow() { return shadow; }
    public int pattern() { return pattern; }
    public boolean monospaced() { return monospaced; }
    public String keyShape() { return keyShape; }
    public String keyMaterial() { return keyMaterial; }
    public double keyOpacity() { return keyOpacity; }
    public String gradientEnd() { return gradientEnd == null ? null : hex(gradientEnd); }
    public boolean gradientHorizontal() { return gradientHorizontal; }
    public double patternOpacity() { return patternOpacity; }
    public String borderColor() { return hex(customBorderColor == null ? accent : customBorderColor); }
    public byte[] photo() { return photo == null ? null : photo.clone(); }
    public double photoShade() { return photoShade; }
    public double photoPosition() { return photoPosition; }
    public String key() {
        return background + ":" + keyBackground + ":" + keyForeground + ":" + accent + ":"
            + actionBackground + ":" + cornerRadius + ":" + borderWidth + ":" + shadow + ":"
            + pattern + ":" + monospaced + ":" + keyShape + ":" + keyMaterial + ":"
            + keyOpacity + ":" + gradientEnd + ":" + gradientHorizontal + ":" + patternOpacity
            + ":" + customBorderColor + ":" + Arrays.hashCode(photo) + ":" + photoShade + ":"
            + photoPosition;
    }

    private static double luminance(int rgb) {
        return .2126 * channel(rgb >> 16) + .7152 * channel(rgb >> 8) + .0722 * channel(rgb);
    }

    private static double channel(int value) {
        double component = (value & 255) / 255.0;
        return component <= .04045 ? component / 12.92 : Math.pow((component + .055) / 1.055, 2.4);
    }
}

File: platforms/harmony/entry/src/main/ets/keyboard/skin/CustomKeyboardSkin.ts
/**
 * Bounded view of the user's custom touch-keyboard design, ported from
 * platforms/android/java/app/msime/client/CustomKeyboardSkin.java.
 *
 * Every field arrives from a shared preference document that the host does not control, so each one
 * is clamped or rejected rather than trusted. Photos are decoded here with a small runtime-neutral
 * Base64 reader, keeping the size and magic-number checks testable without a device image decoder.
 */
const MAX_PHOTO_BYTES: number = 512000;
const KEY_SHAPES: string[] = ['rounded', 'capsule', 'ticket', 'pebble'];
const KEY_MATERIALS: string[] = ['flat', 'raised', 'glass', 'paper'];

export interface CustomSkinDocument {
  readonly background?: number;
  readonly keyBackground?: number;
  readonly keyForeground?: number;
  readonly accent?: number;
  readonly actionBackground?: number;
  readonly cornerRadius?: number;
  readonly borderWidth?: number;
  readonly shadow?: number;
  readonly pattern?: number;
  readonly monospaced?: boolean;
  readonly keyShape?: string;
  readonly keyMaterial?: string;
  readonly keyOpacity?: number;
  readonly gradientEnd?: number | null;
  readonly gradientHorizontal?: boolean;
  readonly patternOpacity?: number;
  readonly customBorderColor?: number | null;
  readonly photo?: string;
  readonly photoShade?: number;
  readonly photoPosition?: number;
}

function bounded(value: number | undefined, minimum: number, maximum: number,
                 fallback: number): number {
  if (value === undefined || !Number.isFinite(value)) {
    return fallback;
  }
  return Math.max(minimum, Math.min(maximum, value));
}

function oneOf(value: string | undefined, allowed: string[]): string {
  return value !== undefined && allowed.includes(value) ? value : allowed[0];
}

function color(value: number | undefined, fallback: number): number {
  return value === undefined || !Number.isFinite(value) ? fallback : (value & 0xffffff);
}

function hex(value: number): string {
  return '#' + (value & 0xffffff).toString(16).toUpperCase().padStart(6, '0');
}

function startsWith(bytes: Uint8Array, prefix: number[]): boolean {
  if (bytes.length < prefix.length) {
    return false;
  }
  for (let index: number = 0; index < prefix.length; index++) {
    if (bytes[index] !== prefix[index]) {
      return false;
    }
  }
  return true;
}

function decodePhoto(value: string | undefined): Uint8Array | null {
  if (value === undefined || value.length === 0 || value.length > 682668
    || value.length % 4 !== 0 || !/^[A-Za-z0-9+/]*={0,2}$/.test(value)) return null;
  const alphabet: string = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/';
  const padding: number = value.endsWith('==') ? 2 : value.endsWith('=') ? 1 : 0;
  const size: number = value.length / 4 * 3 - padding;
  if (size <= 0 || size > MAX_PHOTO_BYTES) return null;
  const output = new Uint8Array(size);
  let cursor: number = 0;
  for (let index: number = 0; index < value.length; index += 4) {
    const a: number = alphabet.indexOf(value[index]);
    const b: number = alphabet.indexOf(value[index + 1]);
    const c: number = value[index + 2] === '=' ? 0 : alphabet.indexOf(value[index + 2]);
    const d: number = value[index + 3] === '=' ? 0 : alphabet.indexOf(value[index + 3]);
    if (a < 0 || b < 0 || c < 0 || d < 0) return null;
    const bits: number = (a << 18) | (b << 12) | (c << 6) | d;
    if (cursor < size) output[cursor++] = (bits >> 16) & 255;
    if (cursor < size) output[cursor++] = (bits >> 8) & 255;
    if (cursor < size) output[cursor++] = bits & 255;
  }
  return output;
}

function photoType(bytes: Uint8Array): string | null {
  if (bytes.length >= 3 && bytes[0] === 0xff && bytes[1] === 0xd8 && bytes[2] === 0xff) {
    return 'image/jpeg';
  }
  if (startsWith(bytes, [0x89, 0x50, 0x4e, 0x47, 13, 10, 26, 10])) return 'image/png';
  if (startsWith(bytes, [0x47, 0x49, 0x46, 0x38])) return 'image/gif';
  if (bytes.length >= 12 && startsWith(bytes, [0x52, 0x49, 0x46, 0x46])
    && bytes[8] === 0x57 && bytes[9] === 0x45 && bytes[10] === 0x42 && bytes[11] === 0x50) {
    return 'image/webp';
  }
  return null;
}

/** JPEG, PNG, GIF and WebP by magic number. Anything else is not drawn. */
export function supportedPhoto(bytes: Uint8Array): boolean {
  if (bytes.length >= 3 && bytes[0] === 0xff && bytes[1] === 0xd8 && bytes[2] === 0xff) {
    return true;
  }
  if (startsWith(bytes, [0x89, 0x50, 0x4e, 0x47, 13, 10, 26, 10])) {
    return true;
  }
  if (startsWith(bytes, [0x47, 0x49, 0x46, 0x38, 0x37, 0x61])
      || startsWith(bytes, [0x47, 0x49, 0x46, 0x38, 0x39, 0x61])) {
    return true;
  }
  return bytes.length >= 12 && startsWith(bytes, [0x52, 0x49, 0x46, 0x46])
    && bytes[8] === 0x57 && bytes[9] === 0x45 && bytes[10] === 0x42 && bytes[11] === 0x50;
}

/** Matches java.util.Arrays.hashCode so the cache key means the same thing on both hosts. */
function photoHash(bytes: Uint8Array | null): number {
  if (bytes === null) {
    return 0;
  }
  let result: number = 1;
  for (let index: number = 0; index < bytes.length; index++) {
    result = (Math.imul(31, result) + bytes[index]) | 0;
  }
  return result;
}

const DEFAULT_BACKGROUND: number = 0xe8f0eb;
const DEFAULT_KEY_BACKGROUND: number = 0xffffff;
const DEFAULT_KEY_FOREGROUND: number = 0x17251d;
const DEFAULT_ACCENT: number = 0x185c47;

/** WCAG relative luminance, which decides whether the action key prints black or white. */
function channel(value: number): number {
  const component: number = (value & 255) / 255;
  return component <= 0.04045 ? component / 12.92 : Math.pow((component + 0.055) / 1.055, 2.4);
}

function luminance(rgb: number): number {
  return 0.2126 * channel(rgb >> 16) + 0.7152 * channel(rgb >> 8) + 0.0722 * channel(rgb);
}

export class CustomKeyboardSkin {
  private backgroundColor: number = DEFAULT_BACKGROUND;
  private keyBackgroundColor: number = DEFAULT_KEY_BACKGROUND;
  private keyForegroundColor: number = DEFAULT_KEY_FOREGROUND;
  private accentColor: number = DEFAULT_ACCENT;
  private actionBackgroundColor: number = DEFAULT_ACCENT;
  private cornerRadiusValue: number = 8;
  private borderWidthValue: number = 0;
  private shadowValue: number = 0;
  private patternValue: number = 0;
  private monospacedValue: boolean = false;
  private keyShapeValue: string = 'rounded';
  private keyMaterialValue: string = 'flat';
  private keyOpacityValue: number = 1;
  private gradientEndColor: number | null = null;
  private gradientHorizontalValue: boolean = false;
  private patternOpacityValue: number = 0.15;
  private customBorderColor: number | null = null;
  private photoBytes: Uint8Array | null = null;
  private photoSourceValue: string | null = null;
  private photoShadeValue: number = 0.25;
  private photoPositionValue: number = 0.5;

  static defaults(): CustomKeyboardSkin {
    return new CustomKeyboardSkin();
  }

  /** `photo` is the already-decoded image; pass null when there is none or it failed to decode. */
  static from(document: CustomSkinDocument | null,
              photo: Uint8Array | null = null): CustomKeyboardSkin {
    const value: CustomKeyboardSkin = CustomKeyboardSkin.defaults();
    if (document === null) {
      return value;
    }
    value.backgroundColor = color(document.background, value.backgroundColor);
    value.keyBackgroundColor = color(document.keyBackground, value.keyBackgroundColor);
    value.keyForegroundColor = color(document.keyForeground, value.keyForegroundColor);
    value.accentColor = color(document.accent, value.accentColor);
    value.actionBackgroundColor = color(document.actionBackground, value.actionBackgroundColor);
    value.cornerRadiusValue = bounded(document.cornerRadius, 0, 20, 8);
    value.borderWidthValue = bounded(document.borderWidth, 0, 2, 0);
    value.shadowValue = bounded(document.shadow, 0, 0.4, 0);
    value.patternValue = Math.max(0, Math.min(3, document.pattern ?? 0));
    value.monospacedValue = document.monospaced ?? false;
    value.keyShapeValue = oneOf(document.keyShape, KEY_SHAPES);
    value.keyMaterialValue = oneOf(document.keyMaterial, KEY_MATERIALS);
    value.keyOpacityValue = bounded(document.keyOpacity, 0.25, 1, 1);
    value.gradientEndColor = document.gradientEnd === undefined || document.gradientEnd === null
      ? null : color(document.gradientEnd, value.backgroundColor);
    value.gradientHorizontalValue = document.gradientHorizontal ?? false;
    value.patternOpacityValue = bounded(document.patternOpacity, 0, 0.5, 0.15);
    value.customBorderColor =
      document.customBorderColor === undefined || document.customBorderColor === null
        ? null : color(document.customBorderColor, value.accentColor);
    value.photoShadeValue = bounded(document.photoShade, 0, 0.8, 0.25);
    value.photoPositionValue = bounded(document.photoPosition, 0, 1, 0.5);
    const decoded: Uint8Array | null = photo ?? decodePhoto(document.photo);
    value.photoBytes = decoded !== null && decoded.length <= MAX_PHOTO_BYTES
      && supportedPhoto(decoded) ? decoded : null;
    const mime: string | null = value.photoBytes === null ? null : photoType(value.photoBytes);
    if (mime !== null && document.photo !== undefined && decoded !== null) {
      value.photoSourceValue = `data:${mime};base64,${document.photo}`;
    }
    return value;
  }

  background(): string { return hex(this.backgroundColor); }
  keyBackground(): string { return hex(this.keyBackgroundColor); }
  keyForeground(): string { return hex(this.keyForegroundColor); }
  accent(): string { return hex(this.accentColor); }
  actionBackground(): string { return hex(this.actionBackgroundColor); }

  /** Black on a light action key, white on a dark one, decided by relative luminance. */
  actionForeground(): string {
    return luminance(this.actionBackgroundColor) > 0.179 ? '#000000' : '#FFFFFF';
  }

  cornerRadius(): number { return this.cornerRadiusValue; }
  borderWidth(): number { return this.borderWidthValue; }
  shadow(): number { return this.shadowValue; }
  pattern(): number { return this.patternValue; }
  monospaced(): boolean { return this.monospacedValue; }
  keyShape(): string { return this.keyShapeValue; }
  keyMaterial(): string { return this.keyMaterialValue; }
  keyOpacity(): number { return this.keyOpacityValue; }
  gradientEnd(): string | null {
    return this.gradientEndColor === null ? null : hex(this.gradientEndColor);
  }
  gradientHorizontal(): boolean { return this.gradientHorizontalValue; }
  patternOpacity(): number { return this.patternOpacityValue; }
  borderColor(): string {
    return hex(this.customBorderColor === null ? this.accentColor : this.customBorderColor);
  }
  photo(): Uint8Array | null { return this.photoBytes; }
  photoSource(): string | null { return this.photoSourceValue; }
  photoShade(): number { return this.photoShadeValue; }
  photoPosition(): number { return this.photoPositionValue; }

  /** Identity for caching a rendered skin. Any field that changes the drawing appears here. */
  key(): string {
    return [
      this.backgroundColor, this.keyBackgroundColor, this.keyForegroundColor, this.accentColor,
      this.actionBackgroundColor, this.cornerRadiusValue, this.borderWidthValue, this.shadowValue,
      this.patternValue, this.monospacedValue, this.keyShapeValue, this.keyMaterialValue,
      this.keyOpacityValue, this.gradientEndColor, this.gradientHorizontalValue,
      this.patternOpacityValue, this.customBorderColor, photoHash(this.photoBytes),
      this.photoShadeValue, this.photoPositionValue
    ].join(':');
  }
}
